river/expert/bandit.py

```diff
@@ -0,0 +1,269 @@
+
+    def learn_one(self, x, y):
+        chosen_arm = self._pull_arm()
+        chosen_model = self[chosen_arm]
+
+        y_pred = chosen_model.predict_one(x)
+        self.metric.update(y_pred=y_pred, y_true=y)
+        chosen_model.learn_one(x=x, y=y)
+
+        # Update bandit internals (common to all bandit)
+        reward = self._compute_scaled_reward(y_pred=y_pred, y_true=y)
+        self._n_iter += 1
+        self._N[chosen_arm] += 1
+        self._average_reward[chosen_arm] += (1.0 / self._N[chosen_arm]) * \
+                                            (reward - self._average_reward[chosen_arm])
+
+        # Specific update of the arm for certain bandit class
+        self._update_arm(chosen_arm, reward)
+
+        if self.print_every:
+            if (self._n_iter % self.print_every) == 0:
+                self._print_info()
+
+        if self.save_percentage_pulled:
+            self.store_percentage_pulled += [self.percentage_pulled]
+
+        if self.save_metric_values:
+            self.metric_values += [self.metric._eval(y_pred, y)]
+
+        return self
+
+    def add_models(self, new_models):
+        if not isinstance(new_models, list):
+            raise TypeError("Argument `new_models` must be of a list")
+
+        length_new_models = len(new_models)
+        # Careful, not validation of the model is done here (contrary to __init__)
+        self.models += new_models
+        self._n_arms += length_new_models
+        self._N = np.concatenate([self._N, np.zeros(length_new_models, dtype=np.int)])
+        self._average_reward = np.concatenate(
+            [self._average_reward, np.zeros(length_new_models, dtype=np.float)]
+        )
+
+    def _compute_scaled_reward(self, y_pred, y_true, update_scaler=True):
+        metric_value = self.metric._eval(y_pred, y_true)
+        metric_to_reward_dict = {
+            "metric": metric_value if self.metric.bigger_is_better else (-1) * metric_value
+        }
+        if update_scaler:
+            self.reward_scaler.learn_one(metric_to_reward_dict)
+        reward = self.reward_scaler.transform_one(metric_to_reward_dict)["metric"]
+        return reward
+
+    def _print_info(self):
+        print(
+            str(self),
+            str(self.metric),
+            "Best model id: " + str(self._best_model_idx),
+            sep="\n\t"
+        )
+
+
+class EpsilonGreedyBandit(Bandit):
+    def __init__(self, epsilon=0.1, epsilon_decay=None, **kwargs):
+        super().__init__(**kwargs)
+        self.epsilon = epsilon
+        self.epsilon_decay = epsilon_decay
+        if epsilon_decay:
+            self._starting_epsilon = epsilon
+        if not self.reward_scaler:
+            self.reward_scaler = preprocessing.StandardScaler()
+
+    def _pull_arm(self):
+        if np.random.rand() > self.epsilon:
+            chosen_arm = np.argmax(self._average_reward)
+        else:
+            chosen_arm = np.random.choice(self._n_arms)
+
+        return chosen_arm
```

Comment:
Again, a lot of this could be implemented in Python, which I suspect would be faster.